### PR TITLE
feat: health check

### LINKED ISSUE #49: @fengjiachun Do we have the same function in java ingester? If not,i think we should implement it.
@fengjiachun Do we have the same function in java ingester? If not,i think we should implement it.

https://github.com/GreptimeTeam/greptime-proto/blob/973f49cde88a582fb65755cc572ebcf6fb93ccf7/proto/greptime/v1/health.proto#L26

_Originally posted by @killme2008 in https://github.com/GreptimeTeam/greptimedb/issues/4674#issuecomment-2339858563_

### PR DESCRIPTION
**Pull request by fengjiachun**
This PR closes #49

### CODE REVIEW

File: ingester-protocol/src/main/java/io/greptime/RouterClient.java [hunk 1]
```diff
@@ -57,19 +62,29 @@ public boolean init(RouterOptions opts) {
...
             this.refresher = REFRESHER_POOL.getObject();
             this.refresher.scheduleWithFixedDelay(
-                    () -> this.router.refresh().whenComplete((r, e) -> {
-                        if (e != null) {
-                            LOG.error("Router cache refresh failed.", e);
-                        } else {
-                            LOG.debug("Router cache refresh {}.", r ? "success" : "failed");
+                    () -> {
+                        try {
+                            Map<Endpoint, Boolean> health = this.checkHealth().get();
```
Comment (fengjiachun) [inline ingester-protocol/src/main/java/io/greptime/RouterClient.java:73]: We must follow the scheduled tasks in order, so we can't make it asynchronous.

Comment (fengjiachun) [reply ingester-protocol/src/main/java/io/greptime/RouterClient.java]: resolved

File: ingester-protocol/src/main/java/io/greptime/RouterClient.java [hunk 2]
```diff
@@ -57,19 +63,41 @@ public boolean init(RouterOptions opts) {
...
+                            if (t != null) {
+                                LOG.warn("Failed to check health", t);
+                                return;
+                            }
+
+                            // I don't want to worry about the overflow issue with long anymore,
+                            // because assuming one increment per second, it will take 292 years
+                            // to overflow. I think that's sufficient.
+                            if (thisOrder < order.get()) {
+                                LOG.warn("Skip outdated health check result, order: {}", thisOrder);
```
Comment (MichaelScofield) [inline ingester-protocol/src/main/java/io/greptime/RouterClient.java:85]: I'm wondering if this "order" comparison is sufficient to guarantee the execution order of check health. Is the following executions possible:

a task is scheduled (task1):

thisOrder = 1 -> `checkHealth()` is called -> the "if" here is passed -> task1 is halted

a new task is scheduled (task2):

thisOrder = 2 -> `checkHealth()` is called -> the "if" here is passed -> the "refresh" below is executed -> task1 is resumed -> the "refresh" below is executed with outdated data, which is not intended.

Comment (fengjiachun) [reply ingester-protocol/src/main/java/io/greptime/RouterClient.java]: Good catch, fixed in https://github.com/GreptimeTeam/greptimedb-ingester-java/pull/50/commits/bc1ddebe5526aa5e0cee66ccad1f8d15cab63b2f

File: ingester-protocol/src/main/java/io/greptime/RouterClient.java [hunk 3]
```diff
@@ -57,19 +65,44 @@ public boolean init(RouterOptions opts) {
...
-                    }),
+                    () -> {
+                        long thisSequence = this.refreshSequencer.incrementAndGet();
+                        checkHealth().whenComplete((r, t) -> {
+                            if (t != null) {
+                                LOG.warn("Failed to check health", t);
+                                return;
+                            }
+
+                            synchronized (this) {
```
Comment (MichaelScofield) [inline ingester-protocol/src/main/java/io/greptime/RouterClient.java:82]: Is this `this` refer to the closure?

Comment (fengjiachun) [reply ingester-protocol/src/main/java/io/greptime/RouterClient.java:82]: this refer to the RouterClient instance

Comment (fengjiachun) [reply ingester-protocol/src/main/java/io/greptime/RouterClient.java]: I changed the lock to `this.refreshSequencer`; it has the same effect but is easier to understand.

**Review by MichaelScofield — APPROVED** [review]

**Pull request merged**